pyro/contrib/gp/models/gpr.py

```diff
@@ -141,3 +142,80 @@ def forward(self, Xnew, full_cov=False, noiseless=True):
             cov = cov + noise
 
         return loc + self.mean_function(Xnew), cov
+
+    def iter_sample(self, noiseless=True):
+        r"""
+        Iteratively constructs a sample from the Gaussian Process posterior.
+
+        Recall that at test input points :math:`X_{new}`, the posterior is
+        multivariate Gaussian distributed with mean and covariance matrix
+        given by :func:`forward`.
+
+        This method samples lazily from this multivariate Gaussian. The advantage
+        of this approach is that later query points can depend upon earlier ones.
+        Particularly useful when the querying is to be done by an optimisation
+        routine.
+
+        .. note:: The noise parameter ``noise`` (:math:`\epsilon`) together with
+            kernel's parameters have been learned from a training procedure (MCMC or
+            SVI).
+
+        :param bool noiseless: A flag to decide if we want to add sampling noise
+            to the samples beyond the noise inherent in the GP posterior.
+        :returns: sampler
+        :rtype: function
+        """
+        # Make these visible in the inner function
+        global X, y, Kff, N
```

Comment:
Instead of using `global` here and below, how about making `sample_next` take these variables as arguments and returning a curried function with `functools.partial`?

```python
def sample_next(X, y, Kff, N, xnew):
    ...

return functools.partial(sample_next, X, y, Kff, xnew)
```

Reply:
The reason I didn't do that was that we change `X` inside the inner function (by writing `X = Xnew`). I thought scoping rules would mean that that change wasn't saved, but honestly I didn't actually try it

Reply:
You can also put these variables into a dictionary and mutate that instead of using `global`